test/parse.jl

```diff
@@ -771,19 +771,26 @@ macro iter()
 end
 end
 let ex = expand(M16096, :(@iter))
-    @test isa(ex, Expr) && ex.head === :body
+    @test isa(ex, Expr) && ex.head === :thunk
 end
 let ex = expand(Main, :($M16096.@iter))
-    @test isa(ex, Expr) && ex.head === :body
+    @test isa(ex, Expr) && ex.head === :thunk
 end
 let ex = expand(@__MODULE__, :(@M16096.iter))
-    @test isa(ex, Expr) && ex.head === :body
+    @test isa(ex, Expr) && ex.head === :thunk
     @test !isdefined(M16096, :foo16096)
-    @test eval(@__MODULE__, ex) === nothing
+    local_foo16096 = eval(@__MODULE__, ex)
+    @test local_foo16096(2.0) == 1
     @test !@isdefined foo16096
-    @test isdefined(M16096, :foo16096)
+    @test !@isdefined it
+    @test !isdefined(M16096, :foo16096)
+    @test !isdefined(M16096, :it)
+    @test_broken typeof(local_foo16096).name.module === M16096
+    @test_broken typeof(local_foo16096).name.mt.module === M16096
+    @test_broken getfield(M16096, typeof(local_foo16096).name.mt.name) === local_foo16096
+    @test_broken !isdefined(@__MODULE__, typeof(local_foo16096).name.mt.name)
```

Comment:
I actually don't think the current behavior is broken. I interpret the purpose of this macro as defining a method in the caller's module. If you consider e.g. the `vectorize` macros, if my module defines `foo` and I write `@vectorize_1arg foo`, I want it to add a new method in my module, not in Base.